---
title: "The Silent Scourge: Investigating Soil-Borne Fungal Diseases in Wheat Crops"
output:
  pdf_document: default
  html_document:
    df_print: paged
date: "2024-01-24"
---

## The Silent Scourge: Investigating Soil-Borne Fungal Diseases in Wheat Crops

The fungi-wheat relationship significantly affects agricultural ecosystems and productivity. Wheat root rot disease, caused by soil-borne fungi, leads to major crop losses. This study explores the link between soil fungal diversity and wheat roots across nine farms to assess the prevalence of these pathogens.

### Qiime2 to Phyloseq

Converting outcomes from Qiime2 data structure to the phyloseq object structure in R using Qiim2R package. Also, the phyloseq package provides functions for organising and manipulating data within the phyloseq object, allowing for diverse analyses such as diversity assessments, differential abundance testing, and taxonomic profile visualisation.

#### Install and load the required packages for the conversion

```{r install, warning=FALSE, message=FALSE}
# Download qiime2R from Github
# if (!requireNamespace("devtools", quietly = TRUE))
#   install.packages("devtools")
# devtools::install_github("jbisanz/qiime2R")
library("qiime2R")

# Download phyloseq from CRAN
# if (!require("BiocManager", quietly = TRUE))
#   install.packages("BiocManager")
# BiocManager::install("phyloseq")
library("phyloseq")
# install.packages("tidyverse")
library("tidyverse")
# install.packages("ggpubr")
library("ggpubr")
```

#### The conversion of Qiime2 to Phyloseq datastructure

```{r qiime2R, warning=FALSE, message=FALSE}
# Convert qiime2 to phyloseq format
physeq.a <- qza_to_phyloseq(
  features = "ITS/table-its-fungi-with-phyla-no-mitochondria-no-chloroplast.qza", # table.qza
  tree = "ITS/phylogeny-align-to-tree-mafft-fasttree/rooted_tree.qza",
  taxonomy = "ITS/taxonomy-fungi.qza",
  metadata = "ITS/meta-table.txt"
)
physeq.a ## confirm the object


## Merge the replicate samples for each Group
physeq.a.group = merge_samples(physeq.a, "Group") # Sum between replicate samples

# (3B) repair factors in the sample metadata
sample_data(physeq.a.group)$Group <- levels(sample_data(physeq.a)$Group)[get_variable(physeq.a.group, "Group")]
sample_data(physeq.a.group)$Soil.Type <- levels(sample_data(physeq.a)$Soil.Type)[get_variable(physeq.a.group, "Soil.Type")]
sample_data(physeq.a.group)$Soil <- levels(sample_data(physeq.a)$Soil)[get_variable(physeq.a.group, "Soil")]
```

#### Assessing ASV distribution (merged and filtered)

We evaluated the number of Amplicon Sequence Variants (ASVs) and visualised the raw read counts per sample within each group. This provided insights into the distribution of ASVs and the sequencing depth of each sample, allowing for a comprehensive assessment of data quality and quantity.

```{r Install_ASV, warning=FALSE, message=FALSE}
# install.packages("ggplot2")
library("ggplot2")
# install.packages("RColorBrewer")
library("RColorBrewer")
```

```{r ASV, warning=FALSE, message=FALSE}

ASV <- sample_sums(physeq.a)
ASV <- as.data.frame(ASV)
ASV$Group <- physeq.a@sam_data$Group

#### pdf
# pdf(file = "Suppl_ASV_ITS.pdf", width = 8,height = 5)
ggplot(ASV, aes(x = Group, y = ASV,  colour = Group))+
  scale_color_brewer(palette = "Paired") +
  geom_point(alpha = 1, position = "jitter", size = 4) + 
  geom_boxplot(alpha = 0, colour = "black", size = 0.8)+ 
  theme_classic() + 
  theme(text = element_text(size=15, colour = "black"), 
        axis.ticks = element_line(colour = "black", size = 1.25),
        axis.line = element_line(colour = 'black', size = 1.25),
        axis.text.x = element_text(angle=45, hjust=1, 
                                   colour = "black", size = 13),
        axis.text.y = element_text(angle=0, hjust=0.5, 
                                   colour = "black",size = 13),
        axis.title.y = element_text(color="black", size=15,face="bold"),
        legend.position = "none")

# Close the PDF device and save the plot to a file
# dev.off()

rm(ASV)
```

#### Beta diversity

Beta diversity is a metric used in microbial community studies to assess species composition dissimilarity between samples, providing insights into microbial community dynamics and their ecological significance. By quantifying variation in community structure and employing visualisation techniques, beta diversity analysis allows comparisons across habitats or treatments, revealing key drivers of community variation and functional implications. It aids our understanding of microbial assemblages and their responses to environmental changes in ecology, environmental science, and human health research.

```{r Beta diversity , message=FALSE, warning=FALSE}
# method options: NMDS / PCoA
NMDS <- ordinate(physeq = physeq.a, method = "NMDS", distance = "bray")

# pdf
# pdf(file = "Fig9A_bata_ITS.pdf", width = 8,height = 5)
# Plot ordination
plot_ordination(physeq = physeq.a,
                      ordination = NMDS,
                      color = "Group",
                      shape = "Group"
) +
  theme_classic() + 
  scale_color_brewer(palette = "Paired") + 
  scale_fill_brewer(palette = "Paired") + 
  geom_point(aes(color = Group), alpha = 1, size = 3) +
  theme(
    text = element_text(size = 18, colour = "black"), 
    axis.ticks = element_line(colour = "black", size = 1.1),
    axis.line = element_line(colour = 'black', size = 1.1),
    axis.text.x = element_text(colour = "black", angle = 0, hjust = 0.5, 
                               size = 13, face = "bold"),
    axis.text.y = element_text(colour = "black", angle = 0, hjust = 0.5, 
                               size = 13, face = "bold"),
    axis.title.y = element_text(color = "black", size = 20, face = "bold"), 
    axis.title.x = element_text(color = "black", size = 20, face = "bold")
  ) + stat_ellipse(geom = "polygon", type="norm", alpha=0.25, aes(fill = Group)) +
  scale_shape_manual(values = c(15, 17, 3, 4, 16, 18, 21, 22, 23)) # Set custom shapes
# Close the PDF device and save the plot to a file
# dev.off()

# Clean up by removing objects that are no longer needed
rm(NMDS)
```

#### Alpha diversity

Alpha diversity, a key ecological concept, denotes species richness within a specific habitat or community. In microbial ecology, it signifies the variety and distribution of microorganisms in a sample or microbiome. It's measured by species richness and evenness, reflecting the number and distribution of species respectively. Alpha diversity is vital for assessing the stability and functional potential of microbial communities and can be influenced by environmental conditions, host factors, and disturbances. Comparing alpha diversity across samples or groups allows researchers to understand the effects of factors like disease or habitat changes on microbial community structure.

Here we use Shannon for alpha diversity.

```{r Alpha diversity, message=FALSE, warning=FALSE}

# available measurements [c("Observed", "Chao1", "ACE", "Shannon", "Simpson", "InvSimpson", "Fisher")]
alpha.object <- cbind(
  x = sample_data(physeq.a),
  y = estimate_richness(physeq.a, measures = 'Shannon')
)

#### pdf
# pdf(file = "Fig9B_Alpha_ITS.pdf", width = 8,height = 5)
# Create a plot for alpha diversity
ggplot(data = alpha.object, aes(x = x.Group, y = Shannon, color = x.Group)) + 
  scale_color_brewer(palette = "Paired") +
  theme_classic() + 
  labs(
    x = element_blank(),                     # No x-axis label
    y = "Alpha Diversity (Shannon)"          # y-axis label
  ) + 
  geom_point(alpha = 1, position = "jitter", size = 4) + 
  geom_boxplot(alpha = 0, colour = "black", size = 0.8)+ 
  theme(
    text = element_text(size = 18, colour = "black"), 
    axis.ticks = element_line(colour = "black", size = 1.1),
    axis.line = element_line(colour = 'black', size = 1.1),
    axis.text.x = element_text(colour = "black", angle = 45, 
                               hjust = 1, size = 13, face = "bold"),
    axis.text.y = element_text(angle = 0, hjust = 0, 
                               colour = "black", size = 13, face = "bold"),
    axis.title.y = element_text(color = "black", size = 15, face = "bold"),
    legend.position = "none"                 # Hide legend
  ) 
# Close the PDF device and save the plot to a file
# dev.off()


# Clean up by removing the alpha.object
rm(alpha.object)
```

#### The proportions of different microbial taxa

Bar plots illustrating relative abundance in microbiome enable to visually represent the proportions of different microbial taxa within a given sample. Each bar corresponds to a specific taxonomic group, with its height indicating the relative abundance of that group in the microbiome.

```{r Install Bar, warning=FALSE, message=FALSE}
# install.packages("plyr")
library("plyr")
```

```{r Bar, warning=FALSE, message=FALSE}
# https://github.com/joey711/phyloseq/issues/901

# Filter taxa with mean abundance > 0.1
physeq2 = filter_taxa(physeq.a.group, function(x) mean(x) > 0.1, TRUE)

# Transform sample counts to relative abundances
physeq3 = transform_sample_counts(physeq2, function(x) x / sum(x))

# Aggregate taxa at the Phylum level
glom <- tax_glom(physeq3, taxrank = 'Phylum')

# Convert the aggregated phyloseq object to a dataframe
data <- psmelt(glom)
# Convert Phylum column to character
data$Phylum <- as.character(data$Phylum)

# Rename phyla with abundance < 1% to "< 1% abund."
data$Phylum[data$Abundance < 0.01] <- "< 1% abund."

# Calculate medians of abundance for each Phylum
medians <- ddply(data, ~Phylum, function(x) c(median = median(x$Abundance)))

# Extract phyla with median abundance <= 0.01
remainder <- medians[medians$median <= 0.01, ]$Phylum

# Rename selected phyla to "Phyla < 1% abund."
data[data$Phylum %in% remainder, ]$Phylum <- "Phyla < 1% abund."

# Rename phyla with abundance < 1% to "Phyla < 1% abund."
data$Phylum[data$Abundance < 0.01] <- "Phyla < 1% abund."

# Plot the data with condensed phyla categories
p <- ggplot(data = data, aes(x = Group, y = Abundance, fill = Phylum))
p + geom_bar(aes(), stat = "identity", position = "stack") +
    scale_color_brewer(palette = "Paired") +
  theme_classic() +   
  theme(text = element_text(size=15, colour = "black"), 
        axis.ticks = element_line(colour = "black", size = 1.25),
        axis.line = element_line(colour = 'black', size = 1.25),
        axis.text.x = element_text(angle=45, hjust=1, 
                                   colour = "black", size = 13),
        axis.text.y = element_text(angle=0, hjust=0.5, 
                                   colour = "black",size = 13),
        axis.title.y = element_text(color="black", size=15,face="bold"), 
        legend.position = "right") 

# Clean up by removing the alpha.object
rm(physeq2, physeq3, glom, data, medians, remainder, p)
```

#### Number of taxa in groups

Examining the cumulative count of taxa within a particular taxonomic group for each distinct subgroup involves the summation of the total number of taxa within that group while eliminating any overlapping taxa based on their names within individual subgroups. This process aids in comprehending the dynamics of microbiome populations.

```{R install_taxa, warning=FALSE, message=FALSE}
library("dplyr")
library("reshape2")
```

```{R num_taxa, warning=FALSE, message=FALSE}
physeq <- merge_samples(physeq.a, "Group", fun = sum)
gentab_levels <- list()
observationThreshold <- 1

# Define the taxonomic levels
genus_levels <- c("Phylum", "Class", "Order", "Family", "Genus", "Species")

# Something wrong with Kingdom Level, removed from the analysis 
# loop through all the taxonomic levels
  for (level in genus_levels) {
    
    # create a factor variable for each level
    genfac <- factor(tax_table(physeq)[, level])
    
    # calculate the abundance of each genus within each sample
    gentab <- apply(otu_table(physeq), MARGIN = 1, function(x) {
      tapply(x, INDEX = genfac, FUN = sum, na.rm = TRUE, simplify = TRUE)
    })
    
    # calculate the number of samples in which each genus is observed above the threshold
    level_counts <- apply(gentab > observationThreshold, 2, sum)
    
    # create a data frame of level counts with genus names as row names
    BB <- as.data.frame(level_counts)
    BB$Group <- row.names(BB)
    
    # add the data frame to the gentab_levels list
    gentab_levels[[level]] <- BB
  }

# Combine all level counts data frames into one data frame
B2 <- gentab_levels %>% reduce(full_join, by = "Group")
  
# Set row names and column names
row.names(B2) <- B2$Group
B2$Group <- NULL
colnames(B2)[1:6] <- genus_levels

print(B2)
# write.csv(B2, file = "ITS/level_counts_by_group.csv", row.names = TRUE)
B2$Group = row.names(B2)
data_long <- melt(B2)
data_long <- melt(B2, id.vars = "Group", 
                  variable.name = "Level", 
                  value.name = "Taxa")


# Plot the data as a line graph using ggplot
# Open a new PDF graphics device
# pdf(file = "Fig7C_line_graph_ITS.pdf", width=8,height=5)
ggplot(data_long, aes(x = Level, y = Taxa, 
                      color = Group, group = Group)) +
  geom_line() +
  geom_point(size = 4) +
  labs(x = "Taxonomic Level", y = "Count", color = "Group") +
  theme_classic() +   scale_color_brewer(palette = "Paired") +
  theme(
    text = element_text(size = 19, colour = "black"), 
    axis.ticks = element_line(colour = "black", size = 1.1),
    axis.line = element_line(colour = 'black', size = 1.5),
    axis.text.x = element_text(colour = "black", angle = 0, hjust = 0.5, 
                               size = 13, face = "bold"),
    axis.text.y = element_text(colour = "black", angle = 0, hjust = 0.5, 
                               size = 13, face = "bold"),
    axis.title.y = element_text(color = "black", size = 14, face = "bold"), 
    axis.title.x = element_text(color = "black", size = 14, face = "bold"),
    legend.title = element_text(size = 13.5, face = "bold"),
    legend.text = element_text(size = 12),
    legend.key.size=unit(0.4,"cm")
  ) +
  scale_x_discrete(guide = guide_axis(n.dodge=2)) +
  scale_y_continuous(breaks=seq(0,400,by=100))


# Close the PDF device and save the plot to a file
# dev.off()  

# Clean up by removing the alpha.object
rm(physeq, gentab_levels, observationThreshold, B2, data_long, genfac, genus_levels, level, level_counts)
```

#### Plotting the top 10 taxa at family level

Generating a visual representation through the plotting of the top 10 taxa at the family level offers a concise overview of the predominant microbial groups within a dataset. This approach is instrumental in identifying and emphasizing significant contributors to the structure and function of microbial communities across diverse groups or conditions.

```{r top10, warning=FALSE, message=FALSE}

## Transform normalised ASVs to proportions 
proportions = transform_sample_counts(physeq.a.group, function(x) 100 * x/sum(x))

##  S
Top10ASVs = names(sort(taxa_sums(proportions), TRUE)[1:13])
Taxtab10 = cbind(tax_table(proportions), Family30 = NA)
Taxtab10[Top10ASVs, "Family30"] <- as(tax_table(proportions)[Top10ASVs, "Family"], "character")
tax_table(proportions) <- tax_table(Taxtab10)

Rgsm10 = prune_taxa(Top10ASVs, proportions)

# plotting
title = "Top 10 - Family level"
plot_bar(Rgsm10, "Group", fill = "Family", title = title) + 
  coord_flip() + 
  ylim(0, 50) +  
  scale_color_brewer(palette = "Paired") +
  theme_classic() +   
  theme(text = element_text(size=15, colour = "black"), 
        axis.ticks = element_line(colour = "black", size = 1.25),
        axis.line = element_line(colour = 'black', size = 1.25),
        axis.text.x = element_text(angle=45, hjust=1, 
                                   colour = "black", size = 13),
        axis.text.y = element_text(angle=0, hjust=0.5, 
                                   colour = "black",size = 13),
        axis.title.y = element_text(color="black", size=15,face="bold"), 
        legend.position = "right", 
        legend.text = element_text(size = 9.5), 
        legend.key.height= unit(0.45, 'cm'),
        legend.key.width= unit(0.45, 'cm')
        )

# Clean up by removing unnecessary objects
rm(proportions, Top10ASVs, Taxtab10, Rgsm10, title)
```

### Upset plot using UpsetR

Venn diagrams are commonly used for visualizing sets, but they can become complex with more than five sets. UpSet graphs, on the other hand, offer a more efficient way to display intersections and complements, especially for larger or multiple datasets. They provide a more intuitive and informative data representation.

```{r install UpSetR, warning=FALSE, message=FALSE}
library("UpSetR")
library("reshape2")
library("plyr")
library("dplyr")
# BiocManager::install("microbiome")
library("microbiome")
```

```{r UpSetR, warning=FALSE, message=FALSE}
# Aggregate taxa at the genus level
B <- aggregate_taxa(physeq.a.group, "Genus", verbose = TRUE)
# Remove undesired genera
# B2 <- subset_taxa(B, !get("Genus") %in% c("uncultured", "Unknown"))

# Remove unwanted taxon names
taxa_to_remove <- c("uncultured", "Unknown")
B2 <- subset_taxa(B, !get("Genus") %in% taxa_to_remove)

# Extract relevant data from the phyloseq object
sample_data <- sample_data(B2)
otu_table <- otu_table(B2)
abundance <- as.vector(otu_table)

# Create a tibble with the extracted data
D <- tibble(
  Sample = rep(sample_data$Group, each = nrow(otu_table)),
  ASV = rep(rownames(otu_table), times = ncol(otu_table)),
  Abundance = abundance
) %>%
  group_by(Sample) %>%
  mutate(rank = rank(desc(Abundance))) %>%
   filter(Abundance > 0) %>%
  ungroup() %>%
  select(Sample, Abundance, ASV)

# Remove the Abundance column
D$Abundance <- NULL

# Rename the second column to "ASV"
names(D)[2] <- "ASV"
names(D)[1] <- "Group"

# Convert data from long to wide format
E <- dcast(D, ASV ~ Group)

# Define a binary function
binary_fun <- function(x) {
  x[is.na(x)] <- 0
  ifelse(x > 0, 1, 0)
}

col = brewer.pal(n = 9, name = "Paired")

# Apply the binary function to columns 2 to 10
temp_df <- apply(E[2:10], 2, binary_fun)
temp_df <- as.data.frame(temp_df)

# Create an UpSet plot
upset_plot <- upset(temp_df, 
                    sets = colnames(temp_df), 
                    sets.bar.color = (col),
                    order.by = "freq", 
                    empty.intersections = "on",
                    mainbar.y.label = "Counts by Pattern of Conditions", 
                    sets.x.label = "Counts by Condition",
                    matrix.color="blue", 
                    mb.ratio = c(0.6, 0.4),
                    point.size= 2.75,
                    line.size = 1.25, 
                    text.scale = 1.5
)


# Open a new PDF graphics device
# pdf(file = "Fig7D_UpSet_ITS.pdf", width=8,height=5)

# Print the UpSet plot
print(upset_plot)

# Close the PDF device and save the plot to a file
# dev.off()  
```

### Phylogenetic tree

Phylogenetic trees are branching diagrams that show the evolutionary relationships between different species or other entities. They are constructed by comparing the organisms' DNA, proteins, or other features. The more similar the features, the more closely related the organisms are thought to be. It can be used to answer how different species are related to each other, when different species diverged from each other, and what are the common ancestors of different species.

Here, we concentrate on *Fusarium*, a large genus of filamentous fungi. These fungi, often referred to as hyphomycetes, are widely found in soil and are associated with plants. While most species are harmless saprobes and form a significant part of the soil microbial community, some can produce mycotoxins in cereal crops that can harm human and animal health if they enter the food chain. Despite the majority of species being harmless, certain Fusarium species and subspecific groups rank among the most critical fungal pathogens of plants and animals.

```{r install.tree, warning=FALSE, message=FALSE}
# Load required libraries
library("phyloseq")   # For handling phylogenetic sequencing data
# BiocManager::install("ggtree")
library("ggtree")     # For tree visualisation
library("scales")     # For scaling transformations
```

```{r tree, warning=FALSE, message=FALSE}
# Load the GlobalPatterns dataset and prune taxa
GP <- prune_taxa(taxa_sums(physeq.a) > 0, physeq.a)  # Remove taxa with zero sums
GP.chl <- subset_taxa(GP, Genus == "Fusarium")   # Subset data based on Phylum value

# Create a ggtree plot
p <- ggtree(GP.chl, ladderize = TRUE) +
  # geom_text2(aes(subset = !isTip, label = label), hjust = -1.5, size = 3.5) +
  geom_tiplab(aes(label = Species), as_ylab=TRUE) +
  geom_point(aes(x = x + hjust, color = Group, 
                 shape = Group, size = Abundance), na.rm = TRUE) +
  scale_size_continuous(trans = log_trans(2)) +
  scale_shape_manual(values = c(15, 17, 3, 4, 16, 18, 21, 22, 23)) + # Set custom
  scale_color_brewer(palette = "Paired") +
  theme(text = element_text(size=13, colour = "black"), 
        axis.ticks = element_line(colour = "black", size = 1.25),
        axis.line = element_line(colour = 'black', size = 1.25) ,
        axis.title.y = element_text(color="black", size=2.5,face="bold"), 
        legend.text = element_text(size = 8),
        legend.key.size=unit(0.4,"cm"),
        legend.title = element_text(size = 10, face = "bold"),
        legend.position = "bottom"
  ) 

# Open a new PDF graphics device
# pdf(file = "Fig10_Tree_ITS.pdf", width=11,height=5)
# Print the ggtree plot
print(p)

# Close the PDF device and save the plot to a file
# dev.off() 
```

From the phylogenetic tree, it was obsered *Fusarium culmorum* and *Fusarium tonkinense* were identified in only one of the samples. However, there was an enrichment of *Fusarium equiseti* , *Fusarium nurragi and Fusarium waltergamsii.*

-   *Fusarium equiseti* is a fungal species and plant pathogen on a varied range of crops. It is considered to be a weak pathogen on cereals and is occasionally found to be associated with 'Fusarium head blight' infected kernels. It is commonly found in tropical and subtropical areas.

-   *Fusarium nurragi* has been isolated from heath species growing on the coasts in the Australian states of Victoria and Western Australia. Although *F. nurragi* has not been assigned to a Fusarium species complex, it is closely related to the *F. heterosporum* and *F. tricinctum* species complexes.

-   *Fusarium waltergamsii* is a fungal species that was first isolated from the roots of wheat plants in Germany. It belongs to the *Fusarium oxysporum* species complex, which comprises many plant pathogens and opportunistic animal pathogens. *Fusarium waltergamsii* can cause wilt and root rot diseases in wheat and other cereals, reducing crop yield and quality.

```{r fungi_2, warning=FALSE, message=FALSE}
# Plotting for the abundance of one specific fungi
# Subset the taxa to Genus from physeq.wheat 
physeq.a.genus <- subset_taxa(physeq.a, Genus == "Fusarium")
physeq.a.equiseti <- subset_taxa(physeq.a, Species == "Fusarium_equiseti")
physeq.a.nurragi <- subset_taxa(physeq.a, Species == "Fusarium_nurragi")
physeq.a.waltergamsii <- subset_taxa(physeq.a, Species == "Fusarium_waltergamsii")

# Calculate the total abundance of Fusarium for each sample
meta = physeq.a@sam_data
otudf = as.data.frame(t(as.data.frame(physeq.a.genus@otu_table)))
meta$Fusarium = rowSums(otudf)
otudf = as.data.frame(t(as.data.frame(physeq.a.equiseti@otu_table)))
meta$F.equiseti = rowSums(otudf)
otudf = as.data.frame(t(as.data.frame(physeq.a.nurragi@otu_table)))
meta$F.nurragi = rowSums(otudf)
otudf = as.data.frame(t(as.data.frame(physeq.a.waltergamsii@otu_table)))
meta$F.waltergamsii = rowSums(otudf)

# Plot a graph of the abundance of Fusarium for each sample grouped by Group:
p1 <- ggplot(subset(meta, Group %in% c("CL.BO","CL.YO",
                                       "CY.BU","CY.YO",
                                       "SC.HE","SC.SH",
                                       "SL.AN","SL.BE",
                                       "SL.SH")),
             aes(x = Group, y = Fusarium,  colour = interaction(Group))) +
  geom_point(alpha = 1, position = "jitter", size = 4) + 
  geom_boxplot(alpha = 0, colour = "black", size = 0.8)+ 
  theme_classic() +   scale_color_brewer(palette = "Paired") +
  labs(x = "", y = "\n Fusarium") +
  theme(text = element_text(size=15, colour = "black"), 
        axis.ticks = element_line(colour = "black", size = 1.25),
        axis.line = element_line(colour = 'black', size = 1.25),
        axis.text.x = element_text(angle=45, hjust=1, 
                                   colour = "black", size = 13),
        axis.text.y = element_text(angle=0, hjust=0.5, 
                                   colour = "black",size = 13),
        axis.title.y = element_text(color="black", size=12,face="bold"), 
        legend.position = "none")

p2 <- ggplot(subset(meta, Group %in% c("CL.BO","CL.YO",
                                       "CY.BU","CY.YO",
                                       "SC.HE","SC.SH",
                                       "SL.AN","SL.BE",
                                       "SL.SH")),
             aes(x = Group, y = F.equiseti,  colour = interaction(Group))) +
  geom_point(alpha = 1, position = "jitter", size = 4) + 
  geom_boxplot(alpha = 0, colour = "black", size = 0.8)+ 
  theme_classic() +   scale_color_brewer(palette = "Paired") +
  labs(x = "", y = "\n F.nurragi") +
  theme(text = element_text(size=15, colour = "black"), 
        axis.ticks = element_line(colour = "black", size = 1.25),
        axis.line = element_line(colour = 'black', size = 1.25),
        axis.text.x = element_text(angle=45, hjust=1, 
                                   colour = "black", size = 13),
        axis.text.y = element_text(angle=0, hjust=0.5, 
                                   colour = "black",size = 13),
        axis.title.y = element_text(color="black", size=12,face="bold"), 
        legend.position = "none")

p3 <- ggplot(subset(meta, Group %in% c("CL.BO","CL.YO",
                                       "CY.BU","CY.YO",
                                       "SC.HE","SC.SH",
                                       "SL.AN","SL.BE",
                                       "SL.SH")),
             aes(x = Group, y = F.nurragi,  colour = interaction(Group))) +
  geom_point(alpha = 1, position = "jitter", size = 4) + 
  geom_boxplot(alpha = 0, colour = "black", size = 0.8)+ 
  theme_classic() +   scale_color_brewer(palette = "Paired") +
  labs(x = "", y = "\n F.nurragi") +
  theme(text = element_text(size=15, colour = "black"), 
        axis.ticks = element_line(colour = "black", size = 1.25),
        axis.line = element_line(colour = 'black', size = 1.25),
        axis.text.x = element_text(angle=45, hjust=1, 
                                   colour = "black", size = 13),
        axis.text.y = element_text(angle=0, hjust=0.5, 
                                   colour = "black",size = 13),
        axis.title.y = element_text(color="black", size=12,face="bold"), 
        legend.position = "none")

p4 <- ggplot(subset(meta, Group %in% c("CL.BO","CL.YO",
                                       "CY.BU","CY.YO",
                                       "SC.HE","SC.SH",
                                       "SL.AN","SL.BE",
                                       "SL.SH")),
             aes(x = Group, y = F.waltergamsii,  colour = interaction(Group))) +
  geom_point(alpha = 1, position = "jitter", size = 4) + 
  geom_boxplot(alpha = 0, colour = "black", size = 0.8)+ 
  theme_classic() +   scale_color_brewer(palette = "Paired") +
  labs(x = "", y = "\n F.waltergamsii") +
  theme(text = element_text(size=15, colour = "black"), 
        axis.ticks = element_line(colour = "black", size = 1.25),
        axis.line = element_line(colour = 'black', size = 1.25),
        axis.text.x = element_text(angle=45, hjust=1, 
                                   colour = "black", size = 13),
        axis.text.y = element_text(angle=0, hjust=0.5, 
                                   colour = "black",size = 13),
        axis.title.y = element_text(color="black", size=12,face="bold"), 
        legend.position = "none") 

# Combine and Arrange the plots
fig <- ggarrange(p1, p2, p3, p4, labels = c("A", "B", "C", "D"), size = 8,
                 ncol = 2, nrow = 2)

# Add labels
fig <- annotate_figure(fig)

# Print the figure
print(fig)

```

```{r sessionInfo, warning=FALSE, message=FALSE}
sessionInfo()
```
