
# UK Crop Microbiome Cryobank
## Agmicrobiomebase

The UK Crop Microbiome Cryobank integrates genomic (DNA) data with a cryobank collection of samples for the soil microbiomes of the UK major crop plant systems. For this project, the microbiomes are from the rhizosphere (the soil surrounding the crop plant roots) and from bulk soil (soil outside the rhizosphere). The Cryobank provides a facility for researchers to source data and samples, including cryo-preserved microbial material and genomic and metagenomic sequences from different soil microbiome environments.

An integrated cryopreserved collection of samples (rhizosphere and bulk soil, bacterial and fungal isolates and DNA) from crop plant systems (barley, oats, oil seed rape, sugar beet and wheat). An open access AgMicrobiomeBase of microbiome data and associated meta-data linked to current public resources such as MGnify.

### Table of contents
--------------------------------
#### A. 16s Amplicon Gene (V3-V4) Sequencing

[01. 16s Sequencing Analysis (markdown)](https://github.com/paytonyau/agmicrobiomebase/blob/main/16s_sequence_process.md)

[02. 16s - Down Stream Analysis (R mark down)](https://github.com/paytonyau/agmicrobiomebase/blob/main/16s_Rmarkdown.Rmd)
[*16s - Down Stream Analysis (pdf format)](https://github.com/paytonyau/agmicrobiomebase/blob/main/16s_Rmarkdown.pdf)

[03. Reference Database Comparison (R markdown)](https://github.com/paytonyau/agmicrobiomebase/blob/main/16s_ref_database_comparison.Rmd) *may need to run it in non R markdown envernment

------------------------------------
#### 16S Documents/Data/Results
[04. DADA2 analysis outcomes (Qiime2 data structure)](https://github.com/paytonyau/agmicrobiomebase/tree/main/16s/%5BQiime2%5DDADA2_outcomes)

[05. Silva [v138] (Qiime2 data structure)](https://github.com/paytonyau/agmicrobiomebase/tree/main/16s/%5BQiime2%5DSilva_138)

[06. GreenGene1 [v13.8] (Qiime2 data structure)](https://github.com/paytonyau/agmicrobiomebase/tree/main/16s/%5BQiime2%5DGreenGenes_13_8)

[07. GreenGenes2 [v2022.10] (Qiime2 data structure)](https://github.com/paytonyau/agmicrobiomebase/tree/main/16s/%5BQiime2%5DGreenGenes2_2022_10)

[08A. Figure 2* Silva [v138] (Qiime2 data structure)](https://github.com/paytonyau/agmicrobiomebase/tree/main/16s/%5BQiime2%5DSilva_138_233_226_3_2_Figure2) *tested for the sequence length distribution

[08B. Figure 2* data(tsv file in zip format)](https://github.com/paytonyau/agmicrobiomebase/tree/main/R-scripts/16s_length_distribution.zip) 

[08C. Figure 2 plotting (R script)](https://github.com/paytonyau/agmicrobiomebase/tree/main/R-scripts/Fig_2_merged_length_distribution.R)

[09. FIGARO (Result)](https://github.com/paytonyau/agmicrobiomebase/tree/main/R-scripts/16s/%5BResult%5DFIGARO)

-----------------------------------------------------------
#### B. ITS Amplicon Gene (ITS1) Sequencing

[09. ITS Sequencing Analysis (markdown)](https://github.com/paytonyau/agmicrobiomebase/blob/main/ITS_sequencing_process.md)

[10. ITS Down Stream Analysis (R markdown)](https://github.com/paytonyau/anti-TNF-resistance/blob/main/09_LPS_Neutrophils_Groups_Comparsions.R)
[*ITS Down Stream Analysis (pdf format)](https://github.com/paytonyau/agmicrobiomebase/blob/main/ITS_Rmarkdown.pdf)

------------------------------------
#### ITS Documents/Data

[11. DADA2 + UNITE (v9.0) analysis outcomes (Qiime2 data structure)](https://github.com/paytonyau/agmicrobiomebase/tree/main/ITS)