---
title: "The UK Crop Microbiome Cryobank - How-to Guide: sequence length selection & taxonomic Levels of reference databases"
author: "Payton Yau"
date: "2024-03-17"
output: md_document
---

# The UK Crop Microbiome Cryobank - How-to Guide: sequence length selection & taxonomic Levels of reference databases


## Trimming and optimal sequence length selection.
We used the longest possible length to find the highest accumulated percentage of the merged sequence length frequency. It assists in trimming out sequences of lower quality and minimises the error rate for the sequence analysis steps in DADA2. This process is part of the events in preprocessing and to prevent “over-merging”.

The denoise setting is for DADA2: 
qiime dada2 denoise-paired \
  -i-demultiplexed-seqs demux.qza \
  --p-trunc-len-f 233 \
  --p-trunc-len-r 226 \
  --p-trim-left-f 0 \
  --p-trim-left-r 0 \
  --p-max-ee-f 3 \
  --p-max-ee-r 2 \
  --p-n-threads 8 \
  --o-representative-sequences rep-seqs.qza \
  --o-table table.qza

The original "16s_length_distribution.RData" extracted from "rep-seqs-with-phyla-no-mitochondria-no-chloroplast.qzv"

```{r install_load.data, warning=FALSE, message=FALSE}
# Load the necessary libraries
library("ggplot2")     
library("scales")     

# Load the merged sequence data
load("16s_length_distribution.RData")
```


### Calculate the length and frequency of each sequence

```{r cal, warning=FALSE, message=FALSE}
# Calculate the length of each sequence
# This step adds a new column to the data frame that contains the length of each sequence
df$SequenceLength <- apply(df, 1, function(x) nchar(x[['Sequence']]))

# Sort the data frame by SequenceLength
# This step sorts the data frame in ascending order of sequence length
df <- df[order(df$SequenceLength), ]

# Calculate the frequency of each sequence length
# This step calculates the frequency of each sequence length and stores the result in a new data frame
df2 <- table(df$SequenceLength)

# Create a data frame for plotting
# This step creates a new data frame that contains the sequence length and its corresponding frequency
plot_df <- data.frame(SequenceLength = as.numeric(names(df2)), Frequency = as.numeric(df2))

# Calculate the accumulated percentage
# This step calculates the accumulated percentage of the sequence length frequency
plot_df$AccumulatedPercentage <- cumsum(plot_df$Frequency) / sum(plot_df$Frequency) * 100

# Filter the data frame
# This step filters the data frame to include only rows where the accumulated percentage is between 0.5% and 99.5%
plot_df2 <- plot_df[plot_df$AccumulatedPercentage >= 0.5 & plot_df$AccumulatedPercentage <= 99.5, ]
```

### Plotting

```{r graph, warning=FALSE, message=FALSE}
# Create a line graph
# This step creates a line graph of the accumulated sequence length frequency
p = ggplot(plot_df2, aes(x = SequenceLength, y = AccumulatedPercentage)) +
  labs(x = "Sequence Length", y = "Accumulated Percentage (%)", 
       title = "Accumulated Sequence Length Frequency (%)") +
  theme_classic() + 
  geom_line(aes(color = Frequency), size = 2) +
  theme(
    text = element_text(size = 15, colour = "black"), 
    axis.ticks = element_line(colour = "black", size = 1.1),
    axis.line = element_line(colour = 'black', size = 1.1),
    axis.text.x = element_text(colour = "black", angle = 0, hjust = 0.5, 
                               size = 13, face = "bold"),
    axis.text.y = element_text(colour = "black", angle = 0, hjust = 0.5, 
                               size = 13, face = "bold"),
    axis.title.y = element_text(color = "black", size = 20, face = "bold"), 
    axis.title.x = element_text(color = "black", size = 20, face = "bold"))

# Add vertical lines and text labels at the 2%, 50%, and 98% points
# This step adds vertical lines and text labels at the 2%, 50%, and 98% points of the accumulated sequence length frequency
p + geom_vline(aes(xintercept = plot_df$SequenceLength[which.min(abs(plot_df$AccumulatedPercentage - 2))]), 
               linetype="dashed", size = 1.5, color = "wheat4") +
  geom_text(aes(x = plot_df$SequenceLength[which.min(abs(plot_df$AccumulatedPercentage - 2))], y = 2, 
                label = paste("2% (", plot_df$SequenceLength[which.min(abs(plot_df$AccumulatedPercentage - 2))], ")", sep = "")), 
            vjust = -6.5, color = "wheat4") +
  geom_vline(aes(xintercept = plot_df$SequenceLength[which.min(abs(plot_df$AccumulatedPercentage - 50))]), 
             linetype="dashed", size = 1.5, color = "forestgreen") +
  geom_text(aes(x = plot_df$SequenceLength[which.min(abs(plot_df$AccumulatedPercentage - 50))], y = 50, 
                label = paste("50% (", plot_df$SequenceLength[which.min(abs(plot_df$AccumulatedPercentage - 50))], ")", sep = "")), 
            vjust = -2, color = "forestgreen") +
  geom_vline(aes(xintercept = plot_df$SequenceLength[which.min(abs(plot_df$AccumulatedPercentage - 98))]), 
             linetype="dashed", size = 1.5, color = "coral4") +
  geom_text(aes(x = plot_df$SequenceLength[which.min(abs(plot_df$AccumulatedPercentage - 98))], y = 98, 
                label = paste("98% (", plot_df$SequenceLength[which.min(abs(plot_df$AccumulatedPercentage - 98))], ")", sep = "")), 
            vjust = 1.5, color = "coral4") +
  scale_x_continuous(breaks = c(400, 405, 410, 415, 420, 425, 430))
```


## Comparative analysis of taxonomic Levels of reference databases

The script provided below is designed to perform a comparative analysis of taxonomic levels of reference databases. 
It operates by counting the number of matched hits, which are then aggregated by names. 
In this process, multiple ASVs with the same names are consolidated into a single unique name. 
This analysis spans across various taxonomic levels, ranging from Kingdom to Species.

```{r install, echo=FALSE}
##### Install required packages
# if (!requireNamespace("devtools", quietly = TRUE)){install.packages("devtools")}
# devtools::install_github("jbisanz/qiime2R") # current version is 0.99.20
library("qiime2R")
library("phyloseq")
library("dplyr")
library("tidyverse")
```

### Data and Path Definitions

```{r prepare list, echo=FALSE}
# Define the taxonomic levels
genus_levels <- c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species")

# Define the datasets and paths
# We compared GreenGene v.1 (13.8), GreenGene v.2 (2022.10), and Silva v.138 reference databases.

dataset_info <- list(
  list(  # GreenGenes 1
    features = "C:/16S/[Qiime2]GreenGenes_13_8/428_228_220_table_gg-13-8-with-phyla-no-mitochondria-no-chloroplast.qza",
    taxonomy = "C:/16S/[Qiime2]GreenGenes_13_8/428_228_220_taxonomy_gg-13-8.qza",
    output_path = "C:/16S/[Qiime2]GreenGenes_13_8/level_counts_by_group_gg1.csv"
  ),
  list(  # GreenGenes2
    features = "C:/16S/[Qiime2]GreenGenes2_2022_10/428_228_220_table_gg_2022_10-with-phyla-no-mitochondria-no-chloroplast.qza",
    taxonomy = "C:/16S/[Qiime2]GreenGenes2_2022_10/428_228_220_taxonomy_gg_2022_10.qza",
    output_path = "C:/16S/[Qiime2]GreenGenes2_2022_10/level_counts_by_group_gg2.csv"
  ),
  list(  # Silva138
    features = "C:/16S/[Qiime2]Silva_138/428_228_220_table_silva138-with-phyla-no-mitochondria-no-chloroplast.qza",
    taxonomy = "C:/16S/[Qiime2]Silva_138/428_228_220_taxonomy_silva138.qza",
    output_path = "C:/16S/[Qiime2]Silva_138/level_counts_by_group_silva138.csv"
  )
)
```

### Data Processing and Analysis

```{r AE, echo=FALSE}
# Loop through each dataset
for (dataset in dataset_info) {
  # Convert qiime2 results to phyloseq format
  physeq <- qza_to_phyloseq(
    features = dataset$features,
    taxonomy = dataset$taxonomy,
    metadata = "C:/16S/16s_meta-table.txt"
  )
  
  physeq.sum <- subset_samples(physeq, Analysis == "Include")
  physeq.sum <- merge_samples(physeq.sum, "Type", fun = sum)
  
  # Create an empty list to store genus-level abundance data for each taxonomic level
  gentab_levels <- list()
  
  # Set observation threshold
  observationThreshold <- 1
  
  # loop through all the taxonomic levels
  for (level in genus_levels) {
    
    # create a factor variable for each level
    genfac <- factor(tax_table(physeq.sum)[, level])
    
    # calculate the abundance of each genus within each sample
    gentab <- apply(otu_table(physeq.sum), MARGIN = 1, function(x) {
      tapply(x, INDEX = genfac, FUN = sum, na.rm = TRUE, simplify = TRUE)
    })
    
    # calculate the number of samples in which each genus is observed above the threshold
    level_counts <- apply(gentab > observationThreshold, 2, sum)
    
    # create a data frame of level counts with genus names as row names
    BB <- as.data.frame(level_counts)
    BB$name <- row.names(BB)
    
    # add the data frame to the gentab_levels list
    gentab_levels[[level]] <- BB
  }
  
  # Combine all level counts data frames into one data frame
  B2 <- gentab_levels %>% reduce(full_join, by = "name")
  
  # Set row names and column names
  row.names(B2) <- B2$name
  B2$name <- NULL
  colnames(B2)[1:7] <- genus_levels
  
  # Write the data frame to a file
  write.csv(B2, file = dataset$output_path, row.names = TRUE)
  
  # Clean up by removing unnecessary objects
  rm(gentab_levels, observationThreshold, BB, B2)
}
```

### Data Visualisation

```{r Visualisation, echo=FALSE}
# Load the reshape2 and ggplot2 libraries
library(reshape2)
library(ggplot2)

GreenGenes.v1 = read.csv("C:/16S/[Qiime2]GreenGenes_13_8/level_counts_by_group_gg1.csv")[5,]
GreenGenes.v2 = read.csv("C:/16S/[Qiime2]GreenGenes2_2022_10/level_counts_by_group_gg2.csv")[5,]
Sliva.v138 = read.csv("C:/16S/[Qiime2]Silva_138/level_counts_by_group_silva138.csv")[5,]

combined_df <- rbind(GreenGenes.v1, GreenGenes.v2, Sliva.v138)
combined_df$X <- c("GreenGenes.v1", "GreenGenes.v2", "Sliva.v138")

data_long <- melt(combined_df, id.vars = "X", variable.name = "Dataset", value.name = "Count")

colnames(data_long) = c("Ref.Database","Taxonomic.Level","Count")

# Convert Taxonomic.Level to a factor and specify the desired order of the levels
data_long$Taxonomic.Level <- factor(data_long$Taxonomic.Level,
                                    levels = c("Kingdom", "Phylum", "Class", "Order", 
                                               "Family", "Genus", "Species"))
```

### Plot the data as a line graph using ggplot

```{r plot, echo=FALSE}

ggplot(data_long, aes(x = Taxonomic.Level, y = Count, color = Ref.Database, group = Ref.Database)) +
  geom_line(size = 2) +
  geom_point(size = 4) +
  scale_color_manual(values = c("Sliva.v138" = "cornflowerblue", 
                                "GreenGenes.v2" = "greenyellow", 
                                "GreenGenes.v1" = "forestgreen")) +
  labs(x = "Taxonomic Level", y = "Count", color = "Reference\nDatabase") +
  theme_classic() + 
  theme(
    text = element_text(size = 19, colour = "black"), 
    axis.ticks = element_line(colour = "black", size = 1.1),
    axis.line = element_line(colour = 'black', size = 1.1),
    axis.text.x = element_text(colour = "black", angle = 0, hjust = 0.5, size = 13, face = "bold"),
    axis.text.y = element_text(colour = "black", angle = 0, hjust = 0.5, size = 13, face = "bold"),
    axis.title.y = element_text(color = "black", size = 14, face = "bold"), 
    axis.title.x = element_text(color = "black", size = 14, face = "bold"),
    legend.title = element_text(size = 13.5, face = "bold"),
    legend.text = element_text(size = 12),
    legend.key.size=unit(0.4,"cm")
  ) +
  scale_x_discrete(guide = guide_axis(n.dodge=2)) +
  scale_y_continuous(breaks=seq(0,1500,by=250))
```

```{r sessionInfo, warning=FALSE, message=FALSE}
sessionInfo()
```